# **`Business Understaing`**

This project aims to clean and analyze United Kingdom's Railway data to derive key insights

The insights the projects aims to derive include

- Trends in Railway use since August 2023

- Determine the most preffered Payment Method

- Compare the performance of different Ticket Classes and Ticket Types and their pricing

- Find the routes with most passengers (Route = Departure station - Arrival Station)

- Compare Railcard and find their frequency of use

- Analogize different purchase types

- Find the major reasons for delays



# **`Data Understanding`**


In [96]:
# Import Necessary Libraries
import pandas as pd
import numpy
import datetime

In [97]:
# Load dataset
uk_railway_data = pd.read_csv(r"..\UK Railway Data\railway.csv")

uk_railway_data.head()

Transaction ID Date of Purchase Time of Purchase Purchase Type  \
0  da8a6ba8-b3dc-4677-b176       2023-12-08         12:41:11        Online   
1  b0cdd1b0-f214-4197-be53       2023-12-16         11:23:01       Station   
2  f3ba7a96-f713-40d9-9629       2023-12-19         19:51:27        Online   
3  b2471f11-4fe7-4c87-8ab4       2023-12-20         23:00:36       Station   
4  2be00b45-0762-485e-a7a3       2023-12-27         18:22:56        Online   

  Payment Method Railcard Ticket Class Ticket Type  Price  \
0    Contactless    Adult     Standard     Advance     43   
1    Credit Card    Adult     Standard     Advance     23   
2    Credit Card      NaN     Standard     Advance      3   
3    Credit Card      NaN     Standard     Advance     13   
4    Contactless      NaN     Standard     Advance     76   

       Departure Station    Arrival Destination Date of Journey  \
0      London Paddington  Liverpool Lime Street      2024-01-01   
1     London Kings Cross                   York      2024-01-01   
2  Liverpool Lime Street  Manchester Piccadilly      2024-01-02   
3      London Paddington                Reading      2024-01-01   
4  Liverpool Lime Street          London Euston      2024-01-01   

  Departure Time Arrival Time Actual Arrival Time Journey Status  \
0       11:00:00     13:30:00            13:30:00        On Time   
1       09:45:00     11:35:00            11:40:00        Delayed   
2       18:15:00     18:45:00            18:45:00        On Time   
3       21:30:00     22:30:00            22:30:00        On Time   
4       16:45:00     19:00:00            19:00:00        On Time   

  Reason for Delay Refund Request  
0              NaN             No  
1   Signal Failure             No  
2              NaN             No  
3              NaN             No  
4              NaN             No

In [98]:
# Check dataset info
uk_railway_data.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Transaction ID       31653 non-null  object
 1   Date of Purchase     31653 non-null  object
 2   Time of Purchase     31653 non-null  object
 3   Purchase Type        31653 non-null  object
 4   Payment Method       31653 non-null  object
 5   Railcard             10735 non-null  object
 6   Ticket Class         31653 non-null  object
 7   Ticket Type          31653 non-null  object
 8   Price                31653 non-null  int64 
 9   Departure Station    31653 non-null  object
 10  Arrival Destination  31653 non-null  object
 11  Date of Journey      31653 non-null  object
 12  Departure Time       31653 non-null  object
 13  Arrival Time         31653 non-null  object
 14  Actual Arrival Time  29773 non-null  object
 15  Journey Status       31653 non-null  object
 16  Reas

## Notes

- The dataset has 31653 rows, and 18 columns of data

- Railcard and Actual Arrive Time columns have missing data

- Reason for delay also has missing values (Could be that this column was updated only when there was a delay)

- Date of Purchase, Time of Purchase, Date of Journey, Departure Time, Arrival Time and Actual Arrival Time	Journey have incorrect datatypes (Object datatype instead of DateTime)


In [99]:
# Check for missing values
missing_values = uk_railway_data.isnull().sum()

missing_values

Transaction ID             0
Date of Purchase           0
Time of Purchase           0
Purchase Type              0
Payment Method             0
Railcard               20918
Ticket Class               0
Ticket Type                0
Price                      0
Departure Station          0
Arrival Destination        0
Date of Journey            0
Departure Time             0
Arrival Time               0
Actual Arrival Time     1880
Journey Status             0
Reason for Delay       27481
Refund Request             0
dtype: int64

In [100]:
# Display basic information about the dataset (categorical columns)
uk_railway_data.describe(include='O')

Transaction ID Date of Purchase Time of Purchase  \
count                     31653            31653            31653   
unique                    31653              128            24351   
top     da8a6ba8-b3dc-4677-b176       2024-02-02         08:16:53   
freq                          1              513                6   

       Purchase Type Payment Method Railcard Ticket Class Ticket Type  \
count          31653          31653    10735        31653       31653   
unique             2              3        3            2           3   
top           Online    Credit Card    Adult     Standard     Advance   
freq           18521          19136     4846        28595       17561   

            Departure Station    Arrival Destination Date of Journey  \
count                   31653                  31653           31653   
unique                     12                     32             121   
top     Manchester Piccadilly  Birmingham New Street      2024-03-09   
freq                     5650                   7742             313   

       Departure Time Arrival Time Actual Arrival Time Journey Status  \
count           31653        31653               29773          31653   
unique             96          203                 623              3   
top          18:45:00     19:05:00            19:05:00        On Time   
freq             2598         1069                1013          27481   

       Reason for Delay Refund Request  
count              4172          31653  
unique                8              2  
top             Weather             No  
freq                995          30535

In [101]:
# Display basic information about the dataset (numerical columns)
uk_railway_data.describe()

Price
count  31653.000000
mean      23.439200
std       29.997628
min        1.000000
25%        5.000000
50%       11.000000
75%       35.000000
max      267.000000

In [102]:
# Check for duplicate rows
duplicate_rows = uk_railway_data.duplicated().sum()

print(f"We have {duplicate_rows} duplicate rows in our dataset.")

We have 0 duplicate rows in our dataset.


In [103]:
# Check for data types of each column
uk_railway_data.dtypes  

Transaction ID         object
Date of Purchase       object
Time of Purchase       object
Purchase Type          object
Payment Method         object
Railcard               object
Ticket Class           object
Ticket Type            object
Price                   int64
Departure Station      object
Arrival Destination    object
Date of Journey        object
Departure Time         object
Arrival Time           object
Actual Arrival Time    object
Journey Status         object
Reason for Delay       object
Refund Request         object
dtype: object

## **Data Cleaning**

In [104]:
# Change Date of Purchase and Date of Journey from Objects to Datetime
uk_railway_data['Date of Purchase'] = pd.to_datetime(uk_railway_data['Date of Purchase'], errors='coerce')
uk_railway_data['Date of Journey'] = pd.to_datetime(uk_railway_data['Date of Journey'], errors='coerce')
uk_railway_data['Date of Purchase'].info()   
uk_railway_data['Date of Journey'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 31653 entries, 0 to 31652
Series name: Date of Purchase
Non-Null Count  Dtype         
--------------  -----         
31653 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 247.4 KB
<class 'pandas.core.series.Series'>
RangeIndex: 31653 entries, 0 to 31652
Series name: Date of Journey
Non-Null Count  Dtype         
--------------  -----         
31653 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 247.4 KB


In [105]:
# Convert columns to datetime64[ns]
cols = [
    'Time of Purchase',
    'Departure Time',
    'Arrival Time',
    'Actual Arrival Time'
]

for col in cols:
    uk_railway_data[col] = pd.to_datetime(
        uk_railway_data[col],
        format='%H:%M:%S',
        errors='coerce'
    )

# Check dtypes
uk_railway_data[cols].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Time of Purchase     31653 non-null  datetime64[ns]
 1   Departure Time       31653 non-null  datetime64[ns]
 2   Arrival Time         31653 non-null  datetime64[ns]
 3   Actual Arrival Time  29773 non-null  datetime64[ns]
dtypes: datetime64[ns](4)
memory usage: 989.3 KB


In [106]:
# Get unique values from all categorical_columns
categorical_columns = uk_railway_data.drop('Transaction ID', axis=1).select_dtypes(include=['object']).columns

for col in categorical_columns:
    unique_values = uk_railway_data[col].unique()
    print(f"Unique values in '{col}' Column: \n{unique_values}\n")


Unique values in 'Purchase Type' Column: 
['Online' 'Station']

Unique values in 'Payment Method' Column: 
['Contactless' 'Credit Card' 'Debit Card']

Unique values in 'Railcard' Column: 
['Adult' nan 'Disabled' 'Senior']

Unique values in 'Ticket Class' Column: 
['Standard' 'First Class']

Unique values in 'Ticket Type' Column: 
['Advance' 'Off-Peak' 'Anytime']

Unique values in 'Departure Station' Column: 
['London Paddington' 'London Kings Cross' 'Liverpool Lime Street'
 'London Euston' 'York' 'Manchester Piccadilly' 'Birmingham New Street'
 'London St Pancras' 'Oxford' 'Reading' 'Edinburgh Waverley'
 'Bristol Temple Meads']

Unique values in 'Arrival Destination' Column: 
['Liverpool Lime Street' 'York' 'Manchester Piccadilly' 'Reading'
 'London Euston' 'Oxford' 'Durham' 'London St Pancras'
 'Birmingham New Street' 'London Paddington' 'Bristol Temple Meads'
 'Tamworth' 'London Waterloo' 'Sheffield' 'Wolverhampton' 'Leeds'
 'Stafford' 'Doncaster' 'Swindon' 'Nottingham' 'Peterborough

## Notes

- Column "Reasons for Delay" has some repetations due to differences in spelling and the use of cases

- We will standardize the wording

In [108]:
# Clean reasons for delay column
uk_railway_data['Reason for Delay'].unique()

array([nan, 'Signal Failure', 'Technical Issue', 'Weather Conditions',
       'Weather', 'Staffing', 'Staff Shortage', 'Signal failure',
       'Traffic'], dtype=object)

In [110]:
# Standardizing the values in the 'Reason for Delay' column for consistency
uk_railway_data['Reason for Delay'] = uk_railway_data['Reason for Delay'].replace({'Signal failure': 'Signal Failure',
                                                                                   'Staff shortage': 'Staff Shortage',
                                                                                    'Weather': 'Weather Conditions',
                                                                                    'Staffing': 'Staff Shortage'})

uk_railway_data['Reason for Delay'].unique()

array([nan, 'Signal Failure', 'Technical Issue', 'Weather Conditions',
       'Staff Shortage', 'Traffic'], dtype=object)